src/sidebar/components/tag-editor.js

```diff
@@ -1,35 +1,197 @@
+   *  the <input>, then add a new tag.
+   */
+  const handleKeyPress = e => {
+    if (e.key === 'Enter' || e.key === ',') {
+      addTag();
+      // preventDefault stops the delimiter from being
+      // added to the input field
+      e.preventDefault();
+    }
+  };
+  /**
+   *  As the user types, change the suggestions query
+   *  based on the value of inputEl.
+   */
+  const handleKeyUp = e => {
+    // Helper to remove any repeated suggestions
+    // that are already tags.
+    const removeDuplicates = suggestions => {
+      const suggestionsSet = [];
+      suggestions.forEach(suggestion => {
+        if (tagList.indexOf(suggestion) < 0) {
+          suggestionsSet.push(suggestion);
+        }
       });
+      return suggestionsSet;
+    };
+
+    if (e.key && e.key !== 'Enter') {
+      // A user typed a letter and we need to filter the results
+      // to update the <dataset> suggestions
+      const suggestions = removeDuplicates(
+        tagsService.filter(inputEl.current.value)
+      );
+      setSuggestions(suggestions);
+    } else if (e.type === 'keyup') {
+      // Assume the user selected an option from <datalist>
+      // either by clicking on it with the mouse or pressing
+      // the 'Enter' key.
+      addTag();
     }
   };
+
+  const suggestionsList = (() => {
+    const suggestionElements = suggestions.map((suggestion, index) => (
+      <option key={index} value={suggestion} />
+    ));
+
+    // Only return a datalist if there is something to render in the list. This
+    // avoid a bug in Firefox where the datalist rendered does not correspond to the
+    // actual list of <option>s in the dom. Wiping out the entire <datalist> tag seems
+    // to force Firefox's hand.
+    if (suggestionElements.length > 0) {
+      return (
+        <datalist
+          id={`tag-editor-datalist-${annotation.id}`}
+          className="tag-editor__suggestions"
+          aria-label="Annotation suggestions"
+        >
+          {suggestionElements}
+        </datalist>
+      );
+    }
+    return null;
+  })();
+
+  return (
+    <section className="tag-editor">
+      <ul
+        className="tag-editor__tag-list"
+        aria-label="Suggested tags for annotation"
+      >
+        {tagList.map(tag => {
+          return (
+            <li
+              key={`${tag}`}
+              className="tag-editor__tag-item"
+              aria-label={`Tag: ${tag}`}
+            >
+              <span className="tag-editor__edit">{tag}</span>
+              <button
+                onClick={() => {
+                  removeTag(tag);
+                }}
+                title={`Remove Tag: ${tag}`}
+                className="tag-editor__delete"
+              >
+                <SvgIcon name="cancel" />
+              </button>
+            </li>
+          );
+        })}
+      </ul>
+      <input
+        list={`tag-editor-datalist-${annotation.id}`}
+        onBlur={handleKeyUp}
```

Comment:
I still think the naming of the event-handling function is a tad funky given that it also handles `blur`s, but it's a minor thing.

Reply:
I think this is left over from previous efforts; I dont think we need to bother with the blur event at all